conbench/entities/distribution.py

```diff
@@ -108,109 +111,160 @@ def get_distribution(benchmark_result, limit):
     )
 
 
-def update_distribution(benchmark_result, limit):
-    from ..db import engine
-
-    if benchmark_result.run.commit.timestamp is None:
-        return
-
-    distribution = get_distribution(benchmark_result, limit).first()
+def update_distribution(
+    benchmark_result: "BenchmarkResult", commit_limit: int
+) -> Optional[dict]:
+    """Try to upsert a Distribution table row, with the given benchmark_result as the
+    most recent BenchmarkResult in the distribution.
+
+    Returns a dict of values if those values were upserted, else returns None. Should
+    not raise anything.
+    """
+    try:
+        distribution_query = _distribution_query(benchmark_result, commit_limit)
+    except CantFindAncestorCommitsError as e:
+        log.debug(
+            f"Not updating distribution: couldn't find ancestor commits, because {e}"
+        )
+        return None
 
+    distribution = distribution_query.first()
     if not distribution:
-        return
+        log.debug("Not updating distribution: the distribution query returned 0 rows")
+        return None
 
     values = dict(distribution)
     hardware_hash = values.pop("hash")
     values["hardware_hash"] = hardware_hash
-    values["limit"] = limit
-
-    with engine.connect() as conn:
-        conn.execute(
-            insert(Distribution.__table__)
-            .values(values)
-            .on_conflict_do_update(
-                index_elements=["case_id", "context_id", "commit_id", "hardware_hash"],
-                set_=values,
-            )
+    values["limit"] = commit_limit
+
+    Session.execute(
+        insert(Distribution.__table__)
+        .values(values)
+        .on_conflict_do_update(
+            index_elements=["case_id", "context_id", "commit_id", "hardware_hash"],
+            set_=values,
         )
-        conn.commit()
+    )
+    Session.commit()
+    return values
 
 
-def get_closest_parent(run):
-    commit = run.commit
-    if commit.timestamp is None:
-        return None
+def get_closest_ancestor(
+    benchmark_result: "BenchmarkResult", branch: Optional[str] = None
+) -> Optional[Commit]:
+    """Given a BenchmarkResult, return the most recent ancestor commit on the given
+    branch that also has a BenchmarkResult in the database with the same
+    hardware/case/context. If branch is not given (default), search all branches.
```

Comment:
> If branch is not given (default), search all branches.

Is the following statement true? Walk the git commit tree upwards and find the first commit that has a benchmark annotation.

How far back does this go at most, up to the origin? If not: maybe "walk at most N steps"?

If my mental model is correct, then in the process of walking backwards we might or might not visit different branches -- do we really care? It seems like "If branch is not given (default)" we do not care.

But if `branch` is given: what really is the behavior then?

It's cool to see all this -- and it makes me wonder how appropriate our database schema / DB choice is for searches like that, and how working with `git` directly (or tooling from the `git` ecosystem that directly operates on git repos in the local file system might help.

Reply:
>Is this guaranteed to be a single commit, implicitly defined by benchmark_result?

Yes, perhaps I should clarify on the docstring: a BenchmarkResult always belongs to one Run which belongs to one Commit. This function attempts to find the closest ancestor commit to the given BenchmarkResult's commit, so that it can look up the distribution's mean and standard deviation at that time, so that we can calculate a z-score.

>Is the following statement true? Walk the git commit tree upwards and find the first commit that has a benchmark annotation.

I'd say "Walk the git commit tree backwards and find the first commit that has a recorded BenchmarkResult that shares the given BenchmarkResult's hardware, case, and context." We've decided in the past that BenchmarkResults with different combos of hardware/case/context are incomparable.

>How far back does this go at most, up to the origin?
Yes, up to the origin. Imposing a commit limit on this query wouldn't improve performance.

>If my mental model is correct, then in the process of walking backwards we might or might not visit different branches
Yes, and this is okay. If branch is not given, we'll return the first matching commit. If the branch is given, we'll return the first matching commit on the given branch. The default behavior of `set_z_scores()` uses this argument to filter to the default branch, so that we only compare the BenchmarkResult to results on the default branch.

Reply:
For Ed's comments:

> this docstring makes it sound like it's going to search any branch

Yeah, I can clarify this docstring. Originally I had two separate functions—`get_closest_ancestor()` and `get_closest_ancestor_default_branch()`—but they were so similar that I decided to combine them. 

>Maybe just a flag for ancestor branch instead of taking any branch?

Since the default branch name isn't standard, we need to figure it out once and pass it to this function.

---

Okay, I just had another thought when writing this. I think this function is redundant with `Run.get_closest_ancestor()`. We could just do `benchmark_result.run.get_closest_ancestor().commit`, as long as I can figure out how to subset that to default branch commits. I'll mess around with this.

Reply:
Okay I deduplicated some code and am now using `Run.get_baseline_run()` in `set_z_scores()`. Got rid of this confusing `get_closest_ancestor()` function entirely.